### PR TITLE
Implement DIP4 integration tests

### PR DESCRIPTION
**Pull request by codablock**
See individual commits. This will later also be used as the basis for quorum commitment merkle root tests (which I'm implementing atm).

### CODE REVIEW

File: qa/rpc-tests/dip4-coinbasemerkleroots.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        expectedDeleted = []
+        expectedUpdated = [new_mn.proTxHash]
+        mnList = self.test_getmnlistdiff(baseBlockHash, self.nodes[0].getbestblockhash(), mnList, expectedDeleted, expectedUpdated)
+        assert(mnList[new_mn.proTxHash].confirmedHash == 0)
+        # Now let the MN get enough confirmations and verify that the MNLISTDIFF now has confirmedHash != 0
+        self.confirm_mns()
+        mnList = self.test_getmnlistdiff(baseBlockHash, self.nodes[0].getbestblockhash(), mnList, expectedDeleted, expectedUpdated)
+        assert(mnList[new_mn.proTxHash].confirmedHash != 0)
+
+        # Spend the coinbase of the previously added MN and test if it appears in "deletedMNs"
```
Comment (UdjinM6) [inline qa/rpc-tests/dip4-coinbasemerkleroots.py]:
```suggestion
        # Spend the collateral of the previously added MN and test if it appears in "deletedMNs"
```

File: qa/rpc-tests/dip4-coinbasemerkleroots.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        hashes = []
+        for mn in sorted(newMNList.values(), key=lambda mn: ser_uint256(mn.proRegTxHash)):
+            hashes.append(hash256(mn.serialize()))
+        merkleRoot = CBlock.get_merkle_root(hashes)
+        assert_equal(merkleRoot, cbtx.merkleRootMNList)
+
+        return newMNList
+
+    def test_getmnlistdiff_base(self, baseBlockHash, blockHash):
+        block = self.nodes[0].getblock(blockHash)
```
Comment (UdjinM6) [inline qa/rpc-tests/dip4-coinbasemerkleroots.py]:
```suggestion
        hexstr = self.nodes[0].getblockheader(blockHash, False)
        header = CBlockHeader()
        header.deserialize(BytesIO(hex_str_to_bytes((hexstr))))
```

File: qa/rpc-tests/dip4-coinbasemerkleroots.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        d = self.test_node.getmnlistdiff(int(baseBlockHash, 16), int(blockHash, 16))
+        assert_equal(d.baseBlockHash, int(baseBlockHash, 16))
+        assert_equal(d.blockHash, int(blockHash, 16))
+
+        # Check that the merkle proof is valid
+        proof = CMerkleBlock(CBlockHeader.from_json(block), d.merkleProof)
```
Comment (UdjinM6) [inline qa/rpc-tests/dip4-coinbasemerkleroots.py]:
```suggestion
        proof = CMerkleBlock(header, d.merkleProof)
```

**Comment by codablock** [comment]
@UdjinM6 applied your changes (squashed into last commit) and also removed https://github.com/dashpay/dash/pull/2832/commits/c0ef7c1738e8ee75285b7efa57083c2bef1995ed as a result of these changes.

File: qa/rpc-tests/dip4-coinbasemerkleroots.py (hunk 1 above)
Comment (UdjinM6) [reply qa/rpc-tests/dip4-coinbasemerkleroots.py]: this one wasn't applied for some reason

**Pull request merged**